src/ocean/application/Daemon.d

```diff
@@ -0,0 +1,814 @@
+        mainAppLogic() in the main Task.
+
+        Params:
+            cl_args = command line args passed to `main`
+
+        Returns:
+            integer success code to operating system
+
+    ***************************************************************************/
+
+    public int run ( istring[] cl_args )
+    {
+        // Default setup of components.
+        this.setupSignals();
+        scope ( exit ) this.signals.clear();
+        this.setupArgs();
+        this.setupUnixSocketCommands();
+
+        // Call pre-args-handling hooks.
+        foreach ( hook; this.hooks.pre_handle_args )
+            hook();
+
+        // Handle command line args and parse config files.
+        this.handleArgs(cl_args);
+        this.parseConfig();
+        this.oneTimeConfiguration();
+
+        // Initialise scheduler, epoll, and epoll clients.
+        initScheduler(this.settings.scheduler_config);
+        this.timers = new Timers; // Needs epoll, so must be constructed here.
+        this.startTimers();
+        this.startSignalHandling();
+        this.unix_socket.startEventHandling(theScheduler.epoll);
+        scope ( exit ) this.unix_socket.shutdown();
+
+        // Lock the PID file, log the version, and start the main app task.
+        this.pid.lock();
+        scope ( exit ) this.pid.unlock();
+        this.logVersion();
+        return TaskScheduler.runInTask(&this.mainAppLogic);
+    }
+
+    /***************************************************************************
+
+        The derived class should implement its main logic here, including config
+        parsing.
+
+        Returns:
+            integer success code to operating system
+
+    ***************************************************************************/
+
+    protected abstract int mainAppLogic ( );
+
+    /***************************************************************************
+
+        Called periodically. The derived class should implement its stats
+        logging here.
+
+    ***************************************************************************/
+
+    protected abstract void onStatsTimer ( );
+
+    /***************************************************************************
+
+        The derived class should implement its signal handling logic here.
+
+        Note that the signals will be handled with a delay of up to a single
+        epoll cycle. This is because the signal handler is synced with the
+        EpollSelectDispatcher instance. This makes it unsuitable to handle
+        critical signals (like `SIGABRT` or `SIGSEGV`) where the application
+        generally shouldn't be allowed to proceed. For these cases, setup an
+        asynchronous signal handler using `sigaction` instead.
+
+        Params:
+            signals = list of signals that fired since the last event loop cycle
+
+    ***************************************************************************/
+
+    protected abstract void onSignal ( int[] signals );
```

Comment:
So no default signal handling at all?

Reply:
Ah good question. This is something that slipped my mind. Thinking about it, though, I think forcing the user to think about how they want to handle signals is perhaps the best way. The default in `DaemonApp` is essentially just one line (`theScheduler.shutdown();`), so I don't think it's especially important. (This could simply be suggested in the docs.) What do you think?

Reply:
I think as it is initialized by default, the scheduler should be shut down by default as well. I would find it inconsistent to shutdown in instance by default that I haven't set up. Also I guess using the daemon base class should be as simple as possible to use especially for new users. Forcing the user to know about these details, ask someone or read docs contradicts that point. Nevertheless providing an application template might be an alternative approach.

Reply:
>I think as it is initialized by default, the scheduler should be shut down by default as well.

I agree with that. The difficulty is that there's no default signal handling, so it's not clear where scheduler shut down should occur.